Unify Scene and Script tabs

Hilderin opened the pull request:
This PR is a work in progress to implement a unification of tabs between scenes and scripts.

Implements a version of the proposal https://github.com/godotengine/godot-proposals/issues/966
I also based these modifications from the discussion on this proposal: https://github.com/godotengine/godot-proposals/issues/4081
There are others proposals on that subject with different approach and suggestion.

The idea was to standardize the Script Editor with what you see in other IDEs by using tabs for Scenes and Scripts. After using Godot for a while, I still struggle every day with the built-in Script Editor. Having a tab over the script that does not correspond to what is under it is too confusing for me after more than 20 years of using tabbed IDEs. Additionally, the navigation between scenes and scripts was always difficult because the Ctrl-Tab and Ctrl-Shift-Tab functionality did not work like other IDEs. It was only working for scenes and did not use the order of the last used scene. I standardized this by adding a popup when using Ctrl-Tab and Ctrl-Shift-Tab, displaying the opened scenes and scripts in one list ordered by last usage.

At the same time, it was important to me to be backward compatible because a lot of users use Godot and like it that way and it's perfect like that! So almost all these modifications are behind a new option in the Editor Settings.

### What’s New
- New option in Editor settings: “Main Editor Tab Content”:
  - Scenes Only: Current/old behavior.
  - Scenes; Scripts and Plugins: New behavior.
- When 'Scenes; Scripts and Plugins' is enabled:
  - When opening a Script, a Help Doc, or the Asset Lib, a tab is created in the same tab container as the scenes.
  - When clicking on a scene tab, the scene is displayed in the 2D/3D view depending on the last view used.
  - When clicking on a script, the script is opened in the script editor.
- **2D/3D button on top**: 
  - When already in 2D/3D view, it only switch to the selected view to allow 2D and 3D on same scene.
  - Otherwise, shows the last edited/selected scene in the selected view.
- **Script button on top**: Opens the last edited script. If no script is edited, it tries to open the script of the edited scene; otherwise, it creates a new script.
- **Script Editor**:
  - The Script list is hidden.
  - **Context menu on a script**:
    - These options from the Script menu are available in the context menu for scripts: Save, Save All, Copy Script Path, Show in File System, Run.
  - **Context menu on a scene**:
    - Added “Copy Scene Path” to standardize with the Script context menu.
- **Changes in both modes**:
  - Tabs navigation: editor/next_tab (Ctrl-Tab) and editor/prev_tab (Ctrl-Shift-Tab) now opens a window to see all your tabs ordered by last usage like VSCode, Visual Studio, or other IDEs.

### Todo

- [x] Keep the order of the tabs on editor restart.
- [ ] Unification of the Main menu and the Script Editor menu. It’s not that bad, but it could be nice to prevent having 2 menus on the screen.
- [ ] Test and manage multi-screen/script editor floating.
- [ ] Test with external script editor.
- [ ] Improve the “empty” scene that is created when all scenes are closed. It’s not natural to have an empty scene created when a scene is closed.
- [ ] Add tooltips on tabs to see the full path. I did not port the setting to show the full path of scripts in tabs.
- [ ] Rename the option ‘Toggle script panel’ in the File menu of the Script Editor when scripts are in tabs.
- [ ] Test plugin with the main screen that handles some type of object or resource.
- [ ] Update `EditorSettings.xml` for the new or updated parameters.
- [ ] Test in distraction-free mode.
- [ ] Test with the parameter `stay_in_script_editor_on_node_selected` inactive.
- [ ] Reopen the Asset Lib on editor restart if it was opened.
- [x] Rename the EditorSetting to be more specific: Updated to `Main Editor Tab Content`

### Questions
- Rename `EditorSceneTabs` to something like `EditorSceneManager`?
- Provide the possibility for plugins to create tabs via `EditorInterface`?
- Port the Script color from the Script list to the tabs?
- Currently, the + button in the tab bar creates a new scene, same as before. Should it create a new scene or script depending on the current editor?


### Some demo

https://github.com/user-attachments/assets/0828c7c1-c832-46c2-b8f3-4325e3b63b2e



Feel free to give me comments and suggestions to improve what was already done.

Hilderin committed:
Script editor as tabs

aaronfranke commented:
How does this work for custom main screens, such as [2.5D](https://github.com/godotengine/godot-demo-projects/tree/master/misc/2.5d)?

shanayyy commented:
I think we should allow tabs to provide different views depending on the file they are editing. This should cover the use case of the main screen menus at the top.

Hilderin commented:
If I understand correctly the question about 2.5D, yes, it manage it. You can still use the button and shortcut to change between 2d and 3d on the current edited scene. When you switch back to this scene, it will reselect the last view you were in.

RedMser commented:
Very excited for this, some initial thoughts specific to your implementation:

1. What does the + button in the tab bar do? IMO it should create a new script when in script editor, otherwise create a new scene like before. Possibly hide when in other editors, until there's a plugin API to extend its behavior.
2. Am I seeing correctly that the last open scene stays open when switching to a script? It's what I would expect, but just wondering that that's the behavior here.
3. If you're not porting the script tab colors, you should also remove the EditorSetting for it.
4. Rename the new EditorSetting. We may want e.g. vertical tab bar or the tab bar at the bottom later on, and that setting should be called tab layout then. Not sure about a better name for your setting, maybe tab bar contents (or usage)?

Hilderin commented:
@shanayyy I'm not sure to understand what you mean exactly, can you explain a bit more?

Hilderin commented:
@RedMser
> 1. What does the + button in the tab bar do? IMO it should create a new script when in script editor, otherwise create a new scene like before. Possibly hide when in other editors, until there's a plugin API to extend its behavior.

Interesting, probably a good idea, I added a question for that in the PR.

> 2. Am I seeing correctly that the last open scene stays open when switching to a script? It's what I would expect, but just wondering that that's the behavior here.

Yes, exactly. The last open scene stays open when script to other type of editor. That keeps the current behavior. Same goes for inspector nodes or resources, etc.

> 3. If you're not porting the script tab colors, you should also remove the EditorSetting for it.

I my opinion, I don't think user should lost functionality. I will probably implement the script color in tabs just for that.

> 4. Rename the new EditorSetting. We may want e.g. vertical tab bar or the tab bar at the bottom later on, and that setting should be called tab layout then. Not sure about a better name for your setting, maybe tab bar contents (or usage)?

Good point, I added a bullet for this in the todo on the PR.

shanayyy commented:
@Hilderin 
Sorry, I just saw the question about custom main screen and got some ideas. 
It's like this, our main editing area (the largest area in the middle) will be open to all resource objects as long as they register their own main editor function. For example: I open a scene resource, the scene resource registers the scene editor function, so the main editing area will open a tab for the scene resource to provide editing. Then I right-click in the inspector to edit a resource object, such as a script, and since it also registers a script editor function, the main editor area will open another tab for script editing. The same logic applies to resource files opened from the FileSystem.

AThousandShips commented on editor/editor_tab.h:
```diff
@@ -0,0 +1,81 @@
+
+	friend class EditorSceneTabs;
+
+private:
```
```suggestion
```

AThousandShips commented on editor/editor_nav_tabs.h:
```diff
@@ -0,0 +1,68 @@
+#ifndef EDITOR_NAV_TABS_H
+#define EDITOR_NAV_TABS_H
+
+#include "editor/editor_tab.h"
```
Forward declare
